### ISSUE TITLE
pg8000 driver does not work (but psycopg2 does)

### ISSUE THREAD

**Issue by seibert** [apache/airflow#161]
Stumbled on this issue with Airflow:  When I use the pg8000 driver with SQLAlchemy, airflow produces the following backtrace when I try to backfill:

```
Traceback (most recent call last):
  File "/home/kts/miniconda/envs/airflow/bin/airflow", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/home/kts/airflow_test/src/airflow/airflow/bin/airflow", line 10, in <module>
    args.func(args)
  File "/home/kts/airflow_test/src/airflow/airflow/bin/cli.py", line 63, in backfill
    ignore_dependencies=args.ignore_dependencies)
  File "/home/kts/airflow_test/src/airflow/airflow/models.py", line 1900, in run
    job.run()
  File "/home/kts/airflow_test/src/airflow/airflow/jobs.py", line 157, in run
    self._execute()
  File "/home/kts/airflow_test/src/airflow/airflow/jobs.py", line 578, in _execute
    elif ti.is_runnable():
  File "/home/kts/airflow_test/src/airflow/airflow/models.py", line 590, in is_runnable
    return self.is_queueable() and not self.pool_full()
  File "/home/kts/airflow_test/src/airflow/airflow/models.py", line 580, in is_queueable
    flag_upstream_failed=flag_upstream_failed)):
  File "/home/kts/airflow_test/src/airflow/airflow/models.py", line 678, in are_dependencies_met
    successes, skipped, done = qry[0]
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2286, in __getitem__
    return list(self[item:item + 1])[0]
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2281, in __getitem__
    return list(res)
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2516, in __iter__
    return self._execute_and_instances(context)
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2531, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 914, in execute
    return meth(self, multiparams, params)
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 323, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1010, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1146, in _execute_context
    context)
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1341, in _handle_dbapi_exception
    exc_info
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 199, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1139, in _execute_context
    context)
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 450, in do_execute
    cursor.execute(statement, parameters)
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/pg8000-1.10.1-py2.7.egg/pg8000/core.py", line 568, in execute
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/pg8000-1.10.1-py2.7.egg/pg8000/core.py", line 1613, in execute
  File "/home/kts/miniconda/envs/airflow/lib/python2.7/site-packages/pg8000-1.10.1-py2.7.egg/pg8000/core.py", line 1761, in handle_messages
sqlalchemy.exc.ProgrammingError: (pg8000.ProgrammingError) ('ERROR', '42883', 'function sum(text) does not exist') [SQL: u'SELECT sum(CASE WHEN (task_instance.state = %s) THEN %s ELSE %s END) AS sum_1, sum(CASE WHEN (task_instance.state = %s) THEN %s ELSE %s END) AS sum_2, count(task_instance.task_id) AS count_1 \nFROM task_instance \nWHERE task_instance.dag_id = %s AND task_instance.task_id IN (%s) AND task_instance.execution_date = %s AND task_instance.state IN (%s, %s, %s, %s) \n LIMIT %s'] [parameters: ('success', 1, 0, 'skipped', 1, 0, 'daily_tck', 'was_market_open_yesterday', datetime.datetime(2015, 7, 19, 0, 0), 'success', 'failed', 'upstream_failed', 'skipped', 1)]
```

The fix is to switch to psycopg2, but I wanted to note this in case someone else had a similar problem with pg8000.

**Comment by mistercrunch** [comment]
It looks like it happens on the subquery that finds the latest task instances for the tasks in a DAG. I'm guessing it's a problem with the SqlAlchemy dialect, but won't dig in since we don't have any Postegres laying around.

You may want to add a note in the installation docs about favoring one driver over the other.

**Comment by mistercrunch** [comment]
Added a note to the installation docs about favoring psycopg2

**Comment by tokenmathguy** [comment]
I was just reviewing this comment and noticed that the traceback listed line numbers for pg8000 core.py that looked a little funny. I then noticed the pg8000 version in the comments is slightly older than what's available in pypi (current is 1.10.2 uploaded in March 2015).

I haven't tried to reproduce the reported issue with the current version of pg8000, but it may be worth re-examination, especially given that pg8000 has a BSD license, is pure python, and may be more amenable to include in some code bases.
